[MSI] Integration tests for MSI source (Web App / Function App / VM)

gladjohn opened the pull request:
Fixes [#3843](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3843)

**Changes proposed in this request**
- Calls into the msi helper service to get environment variables based on resource type
- sets environment variable
- calls the azure resource to get the msi token
- test coverage for Function App, Web App and VM

**Testing**
n/a

**Performance impact**
none

**Documentation**
- [ ] All relevant documentation is updated.

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+            {
+                result = ex.Message;
+            }
+            catch (Exception ex)
```
We should not catch general exceptions here only MsalClientException

gladjohn replied:
resolved

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+
+        public enum AzureResource
+        {
+            webapp,
```
Naming - use UpperCase for enum members. If you need to transfor to string, use ToString().ToLowerInvariant()

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/Infrastructure/HttpProxyManager.cs:
```diff
@@ -0,0 +1,162 @@
+
+            //Add the Authorization header
+            s_httpClient.DefaultRequestHeaders.Authorization = 
+                new AuthenticationHeaderValue("Bearer", token);
```
Cool

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+            AuthenticationResult result = null;
+
+            //Set the Environment Variables
+            bool isEnvironmentVariableSet = await SetEnvironmentVariablesAsync(azureResource)
```
I think what you are doing here is "downloading" all the env variables from the service onto the box. The name should suggest that, e.g. `SetRemoveEnvVariablesAsync` or smth like that.

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+
+        }
+
+        private async Task<bool> SetEnvironmentVariablesAsync(AzureResource resource)
```
Not a fan of returning true / false. C# pattern is to use exceptions, because they convey more information.

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+            }
+            else
+            {
+                return false;
```
If this fails, throw an exception. The exception will propagate and fail the test.

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+                    .WithManagedIdentity(userIdentity)
+                    .ExecuteAsync().ConfigureAwait(false);
+            }
+            catch (Exception ex)
```
Why don't you just let the exception propagate and fail the test?

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+            }
+
+            //Assert
+            Assert.IsNotNull(result.AccessToken);
```
It's never null, you can remove this.

bgavrilMS replied:
How did you resolve this?

gladjohn replied:
sorry missed this, added an alternate test to check for length, so we make sure that the token is always trimmed

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+
+            //Assert
+            Assert.IsNotNull(result.AccessToken);
+            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
```
Can you add an assert on the expiration. Something like expiration is between 0 and 48h or smth? I've seen bugs where the ExpireOn field was set years in advance and we didn't catch it.

gladjohn replied:
added please take a look

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+        private void ClearEnvironmentVariables()
+        {
+            //Clear the environment variables
+            foreach (KeyValuePair<string, string> kvp in s_envVariables)
```
If env variable A is already present on the box, e.g. A = value1 and you download it from the remote, e.g. A = value2, when you call Clear() you'll reset it. But really what you want to do is to set it back to value1.

Have a look at https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/3849/files#diff-057b5a5346431bf0d06e027a642e673ca6292a88765cc00b0e7603dec26b5154R35 - it captures env variables at point X and on Dispose it resets the state to point X.

gladjohn replied:
Fixed all comments except this one, please review the rest. I need to merge main into my branch and work on this, I will hopefully get to this tomorrow. Will update once I am done with it. I also see @neha-bhargava  has merged IMDS, I will also update the tests for IMDS once I get the main

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,189 @@
+        [DataRow(AzureResource.function, UserAssignedClientID, DisplayName = "User Identity Function App")]
+        [DataRow(AzureResource.webapp, Mi_res_id, DisplayName = "ResourceID Web App")]
+        [DataRow(AzureResource.function, Mi_res_id, DisplayName = "ResourceID Function App")]
+        public async Task AcquireMSITokenAsync(AzureResource azureResource, string userIdentity)
```
This is very cool!

gladjohn commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,325 @@
+        [DataTestMethod]
+        [DataRow(MsiAzureResource.WebApp, "", SendMSIHeader.None, DisplayName = "Web App No Header")]
+        [DataRow(MsiAzureResource.WebApp, "", SendMSIHeader.WithWrongValue, DisplayName = "Web App Wrong Header")]
+        public async Task TryAcquireMSITokenHeaderTestAsync(
```
@bgavrilMS @neha-bhargava please take a look at these tests, most of these may not be applicable since I am tweaking the request before it is sent out. But find a few cases where the error message is missing a co-correlation ID or cases where we do not throw the MSI endpoint error message.

gladjohn replied:
I disabled these as they may not be useful for integration, but @neha-bhargava just the https://portal.azure.com token did not work, maybe you can see why?

bgavrilMS replied:
When would we send MSI header with none or wrong value?

gladjohn replied:
we get the value and then the value changes in the web app. Web restarts or MSI endpoint updates can change the header value. We can discuss more when we write the negative cases

neha-bhargava replied:
Should we tweak the request sent out? Will this ever happen?

neha-bhargava replied:
Yes, lets discuss these.

bgavrilMS commented on tests/Microsoft.Identity.Test.Common/TestConstants.cs:
```diff
@@ -460,4 +460,34 @@ internal static class Adfs2019LabConstants
+        ServiceFabric
+    }
+
+    public enum SendMSIHeader
```
It's not clear what these enums do, please add comments.

gladjohn replied:
added

bgavrilMS commented on tests/Microsoft.Identity.Test.Common/TestConstants.cs:
```diff
@@ -460,4 +460,34 @@ internal static class Adfs2019LabConstants
+    public enum SendMSIHeader
+    {
+        Original,
+        WithWrongValue,
```
We have not really discussed negative testing - but in negative scenarios we are: 

1. testing to ensure correct, easy to understand exceptions are thrown back to user (this is unit test territory, try not to use integration tests for these since they are expensive)
2. ensure that the integration point (e.g. MSI service) behaves in a consistent way when hit with "bad" values. 

I'd like to better understand what the negative test strategy is first before reviewing code. Also, to keep PRs reasonable, aim to deliver the smallest piece of code that provides value - the positive tests are enough. Negative tests should be added as separate PR.

gladjohn replied:
removed negative cases

gladjohn commented:
> Please create smaller PR with infra + positive tests and resolve comments. If you do not agree with a comment, please reply or reach out directly to reviewers.

removed negative cases and updated the PR, also there is a separate PR for infra

neha-bhargava commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,199 @@
+        {
+            //Arrange
+            string result = string.Empty;
+            string expectedClientException = "[Managed Identity] Authentication unavailable. " +
```
I am moving the error messages to MsalErrorMessage class so that these can be referred in the tests.

gladjohn replied:
modified the test

neha-bhargava reviewed (approved):
Looks good. Thanks.

pmaytak commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,199 @@
+            }
+
+            //Assert
+            Assert.AreSame(result, expectedClientException);
```
Use [Assert.ThrowsException](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/8e21adc1c5f3c4440044f07b0edcdcc25fe0ba2a/tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/AppServiceTests.cs#L95)

pmaytak commented on tests/Microsoft.Identity.Test.Integration.netfx/Infrastructure/HttpProxyManager.cs:
```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Rename the file to match the class name.

pmaytak commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,199 @@
+                        .WithManagedIdentity(userIdentity)
+                        .ExecuteAsync().ConfigureAwait(false);
+                }
+                catch (MsalClientException ex)
+                {
+                    throw new Exception(ex.Message);
+                }
```
Not needed? Just let it bubble up and break the test?

pmaytak commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,199 @@
+                //Assert
+                //1. MSI Helper service trims the access token, so that the access token from the MSI
+                //   is not fully returned to the calling service or test for security reasons
+                Assert.IsTrue(result.AccessToken.Length == 28);
```
The length requirement seems a bit flaky. Maybe AT returned by the service should just be some constant - "MSI_token" or something?

pmaytak commented on tests/Microsoft.Identity.Test.Integration.netfx/Infrastructure/HttpProxyManager.cs:
```diff
@@ -0,0 +1,162 @@
+    /// In particular, do not change any properties on HttpClient such as BaseAddress, buffer sizes and Timeout. You should
+    /// also not access DefaultRequestHeaders because the getters are not thread safe (use HttpRequestMessage.Headers instead).
+    /// </remarks>
+    internal class ProxyHttpManager : IHttpManager
```
Maybe this could extend HttpManager, to reduce code duplication, and the ExecuteAsync that has logic can be something like protected override.

gladjohn replied:
Yes we can, but the idea is to extend the proxyHttpManager for other http methods as well. that's the reason of implementing iHttpManager.

SameerK-MSFT commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,184 @@
+        public async Task AcquireMSITokenAsync(MsiAzureResource azureResource, string userIdentity)
+        {
+            //Arrange
+
```
nit: extra line

SameerK-MSFT reviewed (approved):
minor extra lines (more than 1 place)

trwalke commented on tests/Microsoft.Identity.Test.LabInfrastructure/LabServiceApi.cs:
```diff
@@ -161,5 +161,17 @@ public async Task<string> GetUserSecretAsync(string lab)
             return JsonConvert.DeserializeObject<LabCredentialResponse>(result).Secret;
         }
+
+        public async Task<string> GetMSIHelperServiceTokenAsync()
```
Isnt this the same token we are already using for other things? maybe it is better to make the token a property or something that can be accessed when needed rather than making a specific method for MSI?

gladjohn replied:
it's the same app but since the service is not hosted under the same lab api, added a different secret so the resources do not share the same secret. This way external teams can use the service with a different secret

neha-bhargava commented on tests/Microsoft.Identity.Test.Common/TestConstants.cs:
```diff
@@ -462,4 +462,15 @@ internal static class Adfs2019LabConstants
     }
+
+    //MSI Azure resources
+    public enum MsiAzureResource
```
In the MSI PR, I have created a class to hold MSI related enums, lets place this enum there as well: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/nebharg%2FMsiApiPrototype/tests/Microsoft.Identity.Test.Common/Core/Helpers/ManagedIdentityTestUtil.cs#L10

gladjohn replied:
I will wait for you to merge and then take this change

neha-bhargava commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,266 @@
+                }).ConfigureAwait(false);
+
+                //Assert
+                Assert.IsTrue(ex.Message.Contains(UserAssignedIdDoesNotExist));
```
Assert error code instead of error message? Since the error message is based on the server response.

gladjohn replied:
this seems to be a very generic error code but the message seems to be specific in this scenario. thoughts? 

![image](https://user-images.githubusercontent.com/90415114/224449415-f672b172-2164-4435-bdc8-317aaa4e1a72.png)

neha-bhargava replied:
I am thinking how will asserting this message help us? Are we going to take some action if the error message changes?

bgavrilMS replied:
`Assert.InformButDontFail` :)

gladjohn replied:
this message is from the Web App Azure Resource when you pass a wrong client id, this is not from the helper service. Helper service does not manipulate any error codes or error messages it simply passes it down

![image](https://user-images.githubusercontent.com/90415114/225173034-9e5e4eba-812b-4f61-8c3f-f5eccfce8cfe.png)

neha-bhargava commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,266 @@
+                }).ConfigureAwait(false);
+
+                //Assert
+                Assert.IsTrue(ex.Message.Contains(WrongScopesErrorMessage));
```
Assert error code.

neha-bhargava commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,266 @@
+        private const string WrongScopesErrorMessage = "An unexpected error occured while fetching the AAD Token";
+
+        //Resource ID of the User Assigned Identity 
+        private const string Mi_res_id = "/subscriptions/c1686c51-b717-4fe0-9af3-24a20a41fb0c/" +
```
nit:
```suggestion
        private const string UamiResourceId = "/subscriptions/c1686c51-b717-4fe0-9af3-24a20a41fb0c/" +
```

neha-bhargava reviewed (approved):
Looks good. Mainly, I would suggest to assert the error code instead of error message since the error message is not a constant but is constructed from response.

gladjohn commented:
> Please create smaller PR with infra + positive tests and resolve comments. If you do not agree with a comment, please reply or reach out directly to reviewers.

@bgavrilMS can you please review the msi helper service?

neha-bhargava commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,301 @@
+
+            IConfidentialClientApplication cca = CreateCCAWithProxy(s_baseURL);
+
+            MsalServiceException ex = await AssertException.TaskThrowsAsync<MsalServiceException>(async () =>
```
I don't think this will be MsalServiceException. When there is no environment variable, IMDS source is created and tries to acquire token from IMDS endpoint which results in HttpRequestException ` System.Net.Http.HttpRequestException: A socket operation was attempted to an unreachable network.`

gladjohn replied:
will need to test this a bit more but I do get a MSALServiceException

gladjohn replied:
modified to HttpRequestException, the problem was going through the proxy. thanks @neha-bhargava

neha-bhargava commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,301 @@
+        [DataTestMethod]
+        [DataRow(MsiAzureResource.Function, NonExistentUserAssignedClientID, DisplayName = "User Identity Function App")]
+        [DataRow(MsiAzureResource.Function, Non_Existent_UamiResourceId, DisplayName = "ResourceID Function App")]
+        public async Task FunctionAppErrorNotInExpectedFormatAsync(MsiAzureResource azureResource, string userIdentity)
```
nit: This can be combined with the test for web app?

gladjohn replied:
just keeping it as a separate error case

neha-bhargava commented on tests/Microsoft.Identity.Test.Integration.netfx/Infrastructure/ProxyHttpManager.cs:
```diff
@@ -0,0 +1,162 @@
+
+        /// <summary>
+        /// Performs the GET request just like <see cref="SendGetAsync(Uri, IDictionary{string, string}, ILoggerAdapter, bool, CancellationToken)"/>
+        /// but does not throw a ServiceUnavailable service exception. Instead, it returns the <see cref="HttpResponse"/> associated
```
nit
```suggestion
        /// but does not throw a ServiceUnavailable exception. Instead, it returns the <see cref="HttpResponse"/> associated
```

neha-bhargava commented on tests/Microsoft.Identity.Test.Integration.netfx/Infrastructure/ProxyHttpManager.cs:
```diff
@@ -0,0 +1,162 @@
+
+        /// <summary>
+        /// Performs the POST request just like <see cref="SendPostAsync(Uri, IDictionary{string, string}, IDictionary{String, String}, ILoggerAdapter, CancellationToken)"/>
+        /// but does not throw a ServiceUnavailable service exception. Instead, it returns the <see cref="HttpResponse"/> associated
```
```suggestion
        /// but does not throw a ServiceUnavailable exception. Instead, it returns the <see cref="HttpResponse"/> associated
```

neha-bhargava commented on tests/Microsoft.Identity.Test.Integration.netfx/Infrastructure/ProxyHttpManager.cs:
```diff
@@ -0,0 +1,162 @@
+
+        /// <summary>
+        /// Performs the POST request just like <see cref="SendPostAsync(Uri, IDictionary{string, string}, HttpContent, ILoggerAdapter, CancellationToken)"/>
+        /// but does not throw a ServiceUnavailable service exception. Instead, it returns the <see cref="HttpResponse"/> associated
```
```suggestion
        /// but does not throw a ServiceUnavailable exception. Instead, it returns the <see cref="HttpResponse"/> associated
```

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,301 @@
+        private const string UserAssignedClientID = "3b57c42c-3201-4295-ae27-d6baec5b7027";
+
+        //Non Existent User Assigned Client ID 
+        private const string NonExistentUserAssignedClientID = "3b57c42c-9999-9999-zz99-d6baec5b7027";
```
Nit: don't use `zz` in there as it is an invalid GUID digit and a potential Guid.Parse in MSAL's logic will break your test. Consider using the Microsoft tenant id - that's a valid invalid clientid :)

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,301 @@
+            //Arrange
+            using (new EnvVariableContext())
+            {
+                //Get the Environment Variables
```
```suggestion
                // Fetch the env variables from the resource and set them locally
```

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,301 @@
+                //2. First token response is from the MSI Endpoint
+                Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+
+                //3. Validate the ExpiresOn falls within a 24 hour range from now
```
Is this true? Don't they cache some tokens as well?

neha-bhargava replied:
No there is no caching on their side.

bgavrilMS commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,301 @@
+                                TimeSpan.FromHours(24));
+
+                result = await cca
+                    .AcquireTokenForClient(s_msi_scopes)
```
You'll need to rebase on top of Neha's changes (they are in main now). No more ConfidentialClient and AcquireTokenForClient

gladjohn commented on tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs:
```diff
@@ -0,0 +1,301 @@
+                SetEnvironmentVariables(envVariables);
+
+                //form the http proxy URI 
+                string uri = s_baseURL + $"MSIToken?" +
```
I did not take this change as there are two endpoints. but I can revisit

gladjohn commented:
build machine seems to have an IMDS endpoint 

![image](https://user-images.githubusercontent.com/90415114/225511549-0e6676e3-d2d8-4132-af27-cc51b147704b.png)
